Applications/Spire/Source/Ui/DropDownMenu2.cpp

```diff
@@ -0,0 +1,250 @@
+#include "Spire/Ui/DropDownMenu2.hpp"
+#include <QFocusEvent>
+#include <QScrollBar>
+#include <QVBoxLayout>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Ui/ScrollArea.hpp"
+
+using namespace boost;
+using namespace boost::signals2;
+using namespace Spire;
+
+namespace {
+  auto BORDER_PADDING() {
+    static auto padding = 2 * scale_width(1);
+    return padding;
+  }
+
+  void adjust_current_index(DropDownMenu2& menu, int amount) {
+    if(menu.count() == 0 || amount == 0) {
+      return;
+    }
+    auto last_index = menu.count() - 1;
+    auto index = [&] {
+      auto index = menu.get_current();
+      if(!index) {
+        if(amount > 0) {
+          return 0;
+        } else {
+          return last_index;
+        }
+      }
+      *index += amount;
+      if(index > last_index) {
+        return 0;
+      } else if(index < 0) {
+        return last_index;
+      }
+      return *index;
+    }();
+    menu.set_current(index);
+  }
+}
+
+DropDownMenu2::DropDownMenu2(std::vector<DropDownItem*> items,
+    QWidget* parent)
+    : QWidget(parent, Qt::Tool | Qt::FramelessWindowHint),
+      m_selected_index(0),
+      m_max_displayed_items(5) {
+  setAttribute(Qt::WA_ShowWithoutActivating);
+  m_shadow = new DropShadow(true, false, this);
+  setObjectName("DropDownMenu");
+  setStyleSheet(R"(
+    #DropDownMenu {
+      background-color: #C8C8C8;
+    })");
+  auto layout = new QVBoxLayout(this);
+  layout->setContentsMargins(scale_width(1), 0, scale_width(1),
+    scale_height(1));
+  m_scroll_area = new ScrollArea(this);
+  m_scroll_area->setFocusProxy(parent);
+  m_scroll_area->setWidgetResizable(true);
+  layout->addWidget(m_scroll_area);
+  auto body = new QWidget(m_scroll_area);
+  m_list_layout = new QVBoxLayout(body);
+  m_list_layout->setContentsMargins({});
+  m_list_layout->setSpacing(0);
+  m_scroll_area->setWidget(body);
+  for(auto& item : items) {
+    m_list_layout->addWidget(item);
+    m_item_hovered_connections.AddConnection(item->connect_highlighted_signal(
+      [=] (const auto& value) { m_hovered_signal(item->get_value()); }));
+    m_item_selected_connections.AddConnection(item->connect_selected_signal(
+      [=] (const auto& value) { on_item_selected(item->get_value(), item); }));
+  }
+  if(m_list_layout->count() > 0) {
+    update_height();
+    static_cast<DropDownItem*>(
+      m_list_layout->itemAt(0)->widget())->set_highlight();
+    m_selected_index = 0;
+    m_current_index = 0;
+    m_selected_signal(get_value(*m_selected_index));
+    m_current_signal(get_value(*m_current_index));
+  }
+  parent->installEventFilter(this);
+  parent->window()->installEventFilter(this);
+  hide();
+}
+
+bool DropDownMenu2::event(QEvent* event) {
+  if(event->type() == QEvent::WindowDeactivate &&
+      focusWidget() != nullptr && !isAncestorOf(focusWidget())) {
+    hide();
+  }
+  return QWidget::event(event);
+}
+
+bool DropDownMenu2::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == parent()) {
```

Comment:
Don't like this, remove it for the time being. It might be reintroduced but it's not clear to me that this is better than having the parent implement this functionality.

Reply:
The original rationale for the DropDownMenu was to avoid having every menu type having to implement that window functionality. We currently have the SecurityInputLineEdit, ComboBox, FilteredDropDownMenu, and the TrendLineStyleMenu that will need to use the DropDownMenu.

I can't think of an implementation off the top of my head, but it would be nice to have a way to seamlessly provide that window functionality to menu inputs.

Reply:
Reuse can be accomplished without breaking the hierarchy. My preference is to avoid breaking some basic principles that get established unless it really has to be. A class that takes as input a parent and a child and moves the child around to follow the parent should be fairly straight forward but nevertheless those options can be considered down the road at the moment that such reuse is needed.

Reply:
The QObject class itself provides the eventFilter(...) function, so it shouldn't be an issue to implement a class that implements that behavior if you want me to do it now. It would look like this: https://gist.github.com/spireken/d723a081fb32eec47606b9e4fc79070e

It needs a better name, but it shouldn't be any more complicated than installing a filter on the parent widget, the parent widget's window, and implementing the move/hide/show behaviors.

The window being moved doesn't generate a local move event for the parent widget, so the parent widget's window will also have to have its events filtered.